otherlibs/runtime_events/runtime_events.ml

```diff
@@ -237,11 +239,43 @@ module User = struct
 
   external user_register : string -> tag option -> 'a Type.t -> 'a t
                          = "caml_runtime_events_user_register"
-  external user_write : 'a t -> 'a -> unit = "caml_runtime_events_user_write"
+  external user_write : bytes -> 'a t -> 'a -> unit = "caml_runtime_events_user_write"
 
   let register name tag typ = user_register name (Some tag) typ
 
-  let write event value = user_write event value
+  let with_write_buffer =
+    (* [caml_runtime_events_user_write] needs a write buffer in which
+       the user-provided serializer will write its data. The user may
+       want to write a lot of custom events fast, so we want to cache
+       the write buffer across calls.
+
+       To be safe for multi-domain programs, we use domain-local
+       storage for the write buffer. To accomodate for multi-threaded
+       programs (without depending on the Thread module), we store
+       a list of caches for each domain. This might leak a bit of
+       memory: the number of buffers for a domain is equal to the
+       maximum number of threads that requested a buffer concurrently,
+       and we never free those buffers. *)
+    let create_buffer () = Bytes.create 1024 in
+    let write_buffer_cache = Domain.DLS.new_key (fun () -> ref []) in
+    fun consumer ->
+      let buffers = Domain.DLS.get write_buffer_cache in
+      let buf =
+        match !buffers with
+        | [] -> create_buffer ()
+        | b::bs -> buffers := bs; b
+      in
+      Fun.protect ~finally:(fun () -> buffers := buf :: !buffers)
+        (fun () -> consumer buf)
```

Comment:
So if I understand correctly your intent:

- This implementation is not vulnerable to #12677 because the worst that can happen is that, if two threads on the domain run `Domain.DLS.get` concurrently, _and_ this is the first time `Domain.DLS.get` is being run on that domain, then they will get different mutable cells; but this will only cause the drop of an allocated cache, which merely causes unnecessary allocation next time, but no functional bug;
- Thread switches may happen on safe points such as the allocations in `create_buffer ()` and `buf :: !buffers`, and as a result elements of the list can be dropped. But likewise, this causes at worst an additional buffer allocation (and we expect it to be rare).

Is it an accurate description of why you think the code is thread-safe?

Reply:
I think we should fix #12677, and proposed #12889 to do so. The code here is not meant to be robust against #12677.

Your point on the fact that `buf :: !buffers` allocates is something that I did not have in mind, thanks! The code is functionally correct for the reason you mention -- it is an the issue with this caching stuff, most ways to get it wrong will still pass the tests. I believe that I could:

1. Pretend that there is no issue and write nothing about that. Very simple!
2. Write a lengthy comment on why dropping buffers is fine here because I don't like it but it never happens in practice. Annoying.
3. Write a CAS loop to push the buffer back, which is more code and less comments.

I went with (3) in a new commit in this PR, could you have a look?

Reply:
In fact I believe that the code with `buffers := buf :: !buffers` is *not* correct because, in addition to something dropping buffers, it could also re-add in the stack some buffers that have been stolen by a thread and not been given back yet. This could result in two thread racing to serialize into the same buffer, which is incorrect.